### PR TITLE
Docs, part 14

### PR DESCRIPTION
**Pull request by arogachev**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | #358, #389, #495

### CODE REVIEW

File: src/Rule/Json.php [hunk 1]
```diff
@@ -15,7 +15,11 @@
 use Yiisoft\Validator\WhenInterface;
 
 /**
- * Validates that the value is a valid json.
+ * Contains a set of options to determine if the value is a valid JSON string.
+ *
+ * @see https://en.wikipedia.org/wiki/JSON
+ * @see https://tools.ietf.org/html/rfc8259
+ * @see BooleanValueHandler Corresponding handler performing the actual validation.
```
Comment (vjik) [inline src/Rule/Json.php:22]:
```suggestion
 * @see JsonHandler Corresponding handler performing the actual validation.
```

File: src/Rule/Subset.php [hunk 2]
```diff
@@ -15,6 +15,16 @@
 use Yiisoft\Validator\WhenInterface;
 
 /**
+ * A variation of {@see In} rule allowing to use the set of values instead of single value as an input for checking if
+ * it's a subset of the set provided in {@see Subset::$values}.
+ *
+ * The order of items in the validated set is not important.
+ *
+ * Nested arrays are supported too in both {@see values} argument and in the validated value (the order of values in
```
Comment (vjik) [inline src/Rule/Subset.php:23]:
```suggestion
 * Nested arrays are supported too in both {@see $values} argument and in the validated value (the order of values in
```

**Review by vjik — CHANGES_REQUESTED** [review]

File: src/Rule/Required.php [hunk 3]
```diff
@@ -25,22 +38,42 @@ final class Required implements RuleWithOptionsInterface, SkipOnErrorInterface,
     use WhenTrait;
 
     /**
-     * @var callable|null
+     * @var callable|null An empty criteria (either a callable or class implementing `__invoke()` method) used to
+     * determine emptiness of the value. The signature must be like the following:
+     *
+     * ```php
+     * function (mixed $value, bool $isAttributeMissing): bool
```
Comment (samdark) [inline src/Rule/Required.php:45]: I'd add extra info about what `$isAttributeMissing` is.

File: src/Rule/RequiredHandler.php [hunk 4]
```diff
@@ -8,22 +8,34 @@
 use Yiisoft\Validator\Exception\UnexpectedRuleException;
 use Yiisoft\Validator\Result;
 use Yiisoft\Validator\RuleHandlerInterface;
+use Yiisoft\Validator\RuleHandlerResolver\SimpleRuleHandlerContainer;
 use Yiisoft\Validator\ValidationContext;
 
 /**
- * Validates that the specified value is passed and not empty.
+ * A handler for {@see Required} rule. Validates that the specified value is passed and not empty.
```
Comment (samdark) [inline src/Rule/RequiredHandler.php:15]:
```suggestion
 * A handler for {@see Required} rule. Validates that the specified value is passed and is not empty.
```

File: src/Rule/RequiredHandler.php [hunk 5]
```diff
@@ -8,22 +8,34 @@
...
  */
 final class RequiredHandler implements RuleHandlerInterface
 {
     /**
-     * @var callable
+     * @var callable An empty criteria (either a callable or class implementing `__invoke()` method) used to
+     * determine emptiness of the value. The signature must be like the following:
+     *
+     * ```php
+     * function (mixed $value, bool $isAttributeMissing): bool
```
Comment (samdark) [inline src/Rule/RequiredHandler.php:26]: isAttributeMissing should be described.

File: src/Rule/Subset.php [hunk 6]
```diff
@@ -15,6 +15,16 @@
 use Yiisoft\Validator\WhenInterface;
 
 /**
+ * A variation of {@see In} rule allowing to use the set of values instead of single value as an input for checking if
+ * it's a subset of the set provided in {@see Subset::$values}.
+ *
+ * The order of items in the validated set is not important.
+ *
+ * Nested arrays are supported too in both {@see $values} argument and in the validated value (the order of values in
```
Comment (samdark) [inline src/Rule/Subset.php:23]:
```suggestion
 * Nested arrays are supported in both {@see $values} argument and in the validated value (the order of values in
```

**Pull request merged**